filesystem/android_super.ksy

```diff
@@ -0,0 +1,174 @@
+meta:
+  id: android_super
+  title: Android Dynamic Partitions metadata
+  application: Android
+  file-extension: img
+  license: CC0-1.0
+  ks-version: 0.8
+  endian: le
+
+doc: |
+  The metadata stored by Android at the beginning of a "super" partition, which
+  is what it calls a disk partition that holds one or more Dynamic Partitions.
+  Dynamic Partitions do more or less the same thing that LVM does on Linux,
+  allowing Android to map ranges of non-contiguous extents to a single logical
+  device. This metadata holds that mapping.
+
+doc-ref: https://source.android.com/devices/tech/ota/dynamic_partitions/implement
+doc-ref: https://android.googlesource.com/platform/system/core/+/master/fs_mgr/liblp/include/liblp/metadata_format.h
+
+seq:
+  - id: padding
+    size: 0x1000
+    type: dummy
+
+  - id: primary_geometry
+    size: 0x1000
+    type: geometry
+
+  - id: backup_geometry
+    type: geometry
+    size: 0x1000
+
+  - id: primary_metadata
+    size: primary_geometry.metadata_max_size
+    type: metadata
+    repeat: expr
+    repeat-expr: primary_geometry.metadata_slot_count
+
+  - id: backup_metadata
+    size: primary_geometry.metadata_max_size
+    type: metadata
+    repeat: expr
+    repeat-expr: primary_geometry.metadata_slot_count
+
+types:
+  dummy: {}
+
+  geometry:
+    seq:
+      - id: magic
+        contents: [0x67, 0x44, 0x6c, 0x61]
+      - id: struct_size
+        type: u4
+      - id: checksum
+        size: 32
+      - id: metadata_max_size
+        type: u4
+      - id: metadata_slot_count
+        type: u4
+      - id: logical_block_size
+        type: u4
+
+  metadata:
+    seq:
+      - id: magic
+        contents: [0x30, 0x50, 0x4c, 0x41]
+      - id: major_version
+        type: u2
+      - id: minor_version
+        type: u2
+      - id: header_size
+        type: u4
+      - id: header_checksum
+        size: 32
+      - id: tables_size
+        type: u4
+      - id: tables_checksum
+        size: 32
+      - id: partitions
+        type: table_descriptor(0)
+      - id: extents
+        type: table_descriptor(1)
+      - id: groups
+        type: table_descriptor(2)
+      - id: block_devices
+        type: table_descriptor(3)
+
+    types:
+      table_descriptor:
+        params:
+          - id: entry_type
+            type: u8
```

Comment:
So these values aren't actually part of the specification at all—they never appear in the serialized data anywhere. I've just created them so that I can parameterize the type that's held inside each `table_descriptor` instance. Given that, do you still think making an enum makes sense? Or is there a way to parameterize `table_descriptor` directly by a type so that I don't need to use numbers at all?

Reply:
> I've just created them so that I can parameterize the type that's held inside each `table_descriptor` instance. Given that, do you still think making an enum makes sense?

Yes, I think so. Defining enums purely for the internal purpose of the KSY implementation is totally OK, and it's commonly used in specs. For example:

https://github.com/kaitai-io/kaitai_struct_formats/blob/ad875ff48d3ec43ae4ddbcf77b98cefe6d3edb7c/macos/resource_compression/dcmp_0.ksy#L85-L94

It's definitely better than repeating plain magic integers everywhere without any obvious meaning.

Reply:
> Or is there a way to parameterize `table_descriptor` directly by a type so that I don't need to use numbers at all?

Unfortunately not yet. This feature idea is tracked in issue https://github.com/kaitai-io/kaitai_struct/issues/135.

Reply:
Cool, makes sense. Should be all set now :)